homeassistant/components/noonlight/__init__.py

```diff
@@ -0,0 +1,195 @@
+        else:
+            if check_api_token.fail_count > 0:
+                persistent_notification.create(
+                        hass,
+                        "Noonlight API token has now been "
+                        "renewed successfully.",
+                        "Noonlight Token Renewal Success",
+                        NOTIFICATION_TOKEN_UPDATE_SUCCESS)
+            check_api_token.fail_count = 0
+
+        async_track_point_in_utc_time(
+            hass, check_api_token, dt_util.utcnow() + next_check_interval)
+
+    check_api_token.fail_count = 0
+
+    @callback
+    def schedule_first_token_check(event):
+        """Schedule the first token renewal when Home Assistant starts up."""
+        async_track_point_in_utc_time(hass, check_api_token, dt_util.utcnow())
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START,
+                               schedule_first_token_check)
+
+    load_platform(hass, 'switch', DOMAIN, None, config)
+
+    return True
+
+
+class NoonlightPlatform():
+    """Platform for interacting with Noonlight from Home Assistant."""
+
+    def __init__(self, hass, conf):
+        """Initialize NoonlightPlatform."""
+        self.hass = hass
+        self.config = conf
+        self._client = None
+        self._time_to_renew = timedelta(hours=2)
+
+    @property
+    def latitude(self):
+        """Return latitude from the Home Assistant configuration."""
+        return self.config \
+            .get(CONF_LATITUDE, self.hass.config.latitude)
+
+    @property
+    def longitude(self):
+        """Return longitude from the Home Assistant configuration."""
+        return self.config \
+            .get(CONF_LONGITUDE, self.hass.config.longitude)
+
+    @property
+    def access_token(self):
+        """Return the access token from the Noonlight Configuration."""
+        return self.config \
+            .get(CONF_ACCESS_TOKEN, {}) \
+            .get('token')
+
+    @property
+    def access_token_expiry(self):
+        """Return the timestamp when the access token expires."""
+        return self.config \
+            .get(CONF_ACCESS_TOKEN, {}) \
+            .get('expires', dt_util.utc_from_timestamp(0))
+
+    @property
+    def access_token_expires_in(self):
+        """Returns the timedelta when the token expires."""
+        return self.access_token_expiry - dt_util.utcnow()
+
+    @property
+    def should_token_be_renewed(self):
+        """Returns true if the token needs to be renewed."""
+        return self.access_token is None \
+            or self.access_token_expires_in <= self._time_to_renew
+
+    @property
+    def client(self):
+        if self._client is None:
+            import noonlight as nl
+            self._client = nl.NoonlightClient(token=self.access_token)
+            api_url = self.config.get(CONF_API_ENDPOINT)
+            self._client._base_url = api_url
+        return self._client
+
+    async def check_api_token(self, force_renew=False):
```

Comment:
The noonlight library only talks to Noonlight's servers, the token is actually refreshed using a Konnected web-service acting as a middle man

Reply:
I don't think we should add this responsibility to home assistant code base. It belongs in a 3rd party library in my opinion.

I'm happy to hear a second opinion from another member.

Reply:
A client library makes sense to me when there's a general purpose use for the library (i.e. can be used in multiple host applications). In this case, Konnected maintains this single endpoint as an "authentication broker" between this specific HA integration and Noonlight's OAuth2 service. The reason for this is that since HA does not run in the cloud, and the code is open source, we need a piece of middleware in the cloud to securely broker the user's Noonlight token via Konnected's partner client_id and secret.

Because this endpoint is not reusable for other Noonlight integrations, I don't think a client library for this is appropriate. The endpoint is maintained specifically for this HA integration so it feels correct that the HA integration is responsible for making that request.

Reply:
If the integration is maintained specifically for HASS and no one else can utilize it, I’m okay with it living in the HASS codebase. You all will need to be extra diligent to own and maintain the code going forward, however; if this points at a Konnected middleware, others in the community won’t be able to easily troubleshoot or help support.